src/ReturnController.php

```diff
@@ -169,4 +197,56 @@ public function rest_api_adyen_return( WP_REST_Request $request ) {
 		 */
 		return new \WP_REST_Response( null, 303, [ 'Location' => $payment->get_return_redirect_url() ] );
 	}
+
+	/**
+	 * REST API Adyen redirect handler.
+	 *
+	 * @param WP_REST_Request $request Request.
+	 * @return object
+	 */
+	public function rest_api_adyen_redirect( WP_REST_Request $request ) {
+		$payment_id = $request->get_param( 'payment_id' );
+
+		// Payment ID.
+		if ( null === $payment_id ) {
+			return new WP_Error(
+				'pronamic-pay-adyen-no-payment-id',
+				__( 'No payment ID given in `payment_id` parameter.', 'pronamic_ideal' )
+			);
+		}
+
+		$payment = \get_pronamic_payment( $payment_id );
+
+		if ( null === $payment ) {
+			return new WP_Error(
+				'pronamic-pay-adyen-payment-not-found',
+				sprintf(
+					/* translators: %s: payment ID */
+					__( 'Could not find payment with ID `%s`.', 'pronamic_ideal' ),
+					$payment_id
+				),
+				$payment_id
+			);
+		}
+
+		/**
+		 * Result code.
+		 * 
+		 * @link https://docs.adyen.com/online-payments/payment-result-codes
+		 */
+		$result_code = $request->get_param( 'resultCode' );
+
+		$status = ResultCode::transform( $result_code );
+
+		if ( null !== $status ) {
+			$payment->set_status( $status );
+		}
```

Comment:
There's a `$payment->save()` missing, which prevents the payment status from being updated.

Apart from that, I don't think it's really safe to update the payment status this way. The redirect URL can be found in the HTML source of the drop-in checkout page. Simply adding `?resultCode=Authorised` to the URL will update the payment status as if the payment has been completed successfully, without actually paying anything.

- https://example.com/wp-json/pronamic-pay/adyen/v1/redirect/40849?resultCode=Authorised

Can we improve this with for example a hash?

Reply:
I omitted the `$payment->save()` for that reason, Adyen advices to await the webhook call for status updates.

Reply:
Aaaaaah, I understand. Perfect 👌